src/Utilities/CssDocument.php

```diff
@@ -110,175 +113,62 @@ public function getStyleRulesData(array $allowedMediaTypes): array
-     * others will be extracted by this method.
-     *
-     * These rules can be placed anywhere in the CSS and are not case sensitive.
-     *
-     * `@font-face` rules will be checked for validity, though other at-rules will be assumed to be valid.
-     *
-     * @param string $css CSS with comments, `@import` and `@charset` removed
-     *
-     * @return array{0: string, 1: string}
-     *         The first element is the CSS with the at-rules removed.  The second element contains a concatenation of
-     *         the valid at-rules, each followed by whatever whitespace followed it in the original CSS (so that either
-     *         unminified or minified formatting is preserved); if there were no at-rules, it will be an empty string.
-     */
-    private function extractNonConditionalAtRules(string $css): array
-    {
-        $possiblyModifiedCss = $css;
-        $atRules = [];
-
-        while (\preg_match(self::NON_CONDITIONAL_AT_RULE_PATTERN, $possiblyModifiedCss, $matches)) {
-            /** @var array<int, string> $matches */
-            [$fullMatch, $atRuleName] = $matches;
+        $atRulesDocument = new SabberwormCssDocument();
+        $atRulesDocument->setContents($atRules);
 
-            if ($this->isValidAtRule($atRuleName, $fullMatch)) {
-                $atRules[] = $fullMatch;
-            }
-
-            $possiblyModifiedCss = \str_replace($fullMatch, '', $possiblyModifiedCss);
-        }
-
-        return [$possiblyModifiedCss, \implode('', $atRules)];
+        /** @var string $renderedRules */
+        $renderedRules = $atRulesDocument->render();
+        return $renderedRules;
     }
 
     /**
-     * Tests if an at-rule is valid.  Currently only `@font-face` rules are checked for validity; others are assumed to
-     * be valid.
+     * Tests if a CSS rule is an at-rule that should be passed though and copied to a `<style>` element unmodified:
+     * - `@charset` rules are discarded - only UTF-8 is supported - `false` is returned;
+     * - `@import` rules are passed through only if they satisfy the specification ("user agents must ignore any
+     *   '@import' rule that occurs inside a block or after any non-ignored statement other than an '@charset' or an
+     *   '@import' rule");
+     * - `@media` rules are processed separately to see if their nested rules apply - `false` is returned;
+     * - `@font-face` rules are checked for validity - they must contain both a `src` and `font-family` property;
+     * - other at-rules are assumed to be valid and treated as a black box - `true` is returned.
      *
-     * @param string $atIdentifier name of the at-rule with the preceding at sign
-     * @param string $rule full content of the rule, including the at-identifier
+     * @param CssRenderable $rule
      *
      * @return bool
      */
-    private function isValidAtRule(string $atIdentifier, string $rule): bool
+    private function isValidAtRuleToRender(CssRenderable $rule): bool
```

Comment:
Would a decicated method in sabberworm/php-css-parser make sense for this (i.e., should we open a ticket)?

Reply:
No, this filtering is quite specific to our use case.

Here, we discard `@media` rules because they have been processed separately (so that their ordering with respect to other uninlinable CSS is correct, and they can be discarded if they don't match anything).  We could in theory process other conditional at-rules (e.g. #1078) in the same way.

We deliberately discard `@charset` rules, which should otherwise be passed on, because we only support UTF-8.

We have sort-of decided (see #870) to discard `@font-face` rules without at least a `src` and `font-family`, which is the minimum to be meaningful.

We are following the user-agent spec and discarding invalidly-placed `@import` rules (I think it's a "user agents MUST ignore `@import` rules which are preceded by a non-ignored statement other than an `@charset` rule or another `@import` rule" [from memory!]).  This I think is the only aspect where a different, alternative or enhanced behaviour of the Sabberworm parser might be useful more widely.